presto-main/src/main/java/com/facebook/presto/operator/StreamingAggregationOperator.java

```diff
@@ -0,0 +1,298 @@
+    private final PageBuilder pageBuilder;
+    private Page pendingInput;
+    private Page previousPage;
+    private boolean hasOutput;
+    private boolean finishing;
+
+    public StreamingAggregationOperator(OperatorContext operatorContext, List<Type> groupByTypes, List<Integer> groupByChannels, Step step, List<AccumulatorFactory> accumulatorFactories, DataSize minOutputPageSize, int minOutputPageRowCount)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.systemMemoryContext = operatorContext.newLocalSystemMemoryContext();
+        this.userMemoryContext = operatorContext.localUserMemoryContext();
+        this.groupByTypes = ImmutableList.copyOf(requireNonNull(groupByTypes, "groupByChannels is null"));
+        this.groupByChannels = ImmutableList.copyOf(requireNonNull(groupByChannels, "groupByChannels is null"));
+        this.accumulatorFactories = requireNonNull(accumulatorFactories, "accumulatorFactories is null");
+        this.step = requireNonNull(step, "step is null");
+
+        this.aggregates = setupAggregates(step, accumulatorFactories);
+        this.pageBuilder = new PageBuilder(toTypes(groupByTypes, aggregates));
+        this.minOutputPageSize = minOutputPageSize;
+        this.minOutputPageRowCount = minOutputPageRowCount;
+    }
+
+    private List<Aggregator> setupAggregates(Step step, List<AccumulatorFactory> accumulatorFactories)
+    {
+        ImmutableList.Builder<Aggregator> builder = ImmutableList.builder();
+        for (AccumulatorFactory factory : accumulatorFactories) {
+            builder.add(new Aggregator(factory, step));
+        }
+        return builder.build();
+    }
+
+    private static List<Type> toTypes(List<Type> groupByTypes, List<Aggregator> aggregates)
+    {
+        ImmutableList.Builder<Type> builder = ImmutableList.builder();
+        builder.addAll(groupByTypes);
+        aggregates.stream().forEach(aggregator -> builder.add(aggregator.getType()));
+        return builder.build();
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finishing && pendingInput == null;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        checkState(pendingInput == null);
+        requireNonNull(page, "page is null");
+
+        processInput(page);
+        updateMemoryUsage();
+    }
+
+    private void updateMemoryUsage()
+    {
+        long memorySize = pageBuilder.getRetainedSizeInBytes();
+        for (Aggregator aggregator : aggregates) {
+            memorySize += aggregator.getEstimatedSize();
+        }
+
+        if (step.isOutputPartial()) {
+            systemMemoryContext.setBytes(memorySize);
+        }
+        else {
+            userMemoryContext.setBytes(memorySize);
+        }
+    }
+
+    private void processInput(Page page)
```

Comment:
In this snipped I used row by row iteration, but that is not how `Accumulator` work. We should operate on ranges, but principals stay the same.

Reply:
@sopel39 Karol, thanks for the suggestion. I used some of the ideas to simplify the logic, but couldn't get it to quite the shape you asked for. Any chance you could take another look?

Reply:
I think the logic in this class (and this method in particular) would be easier to follow if instead of tracking "previousRow" you just kept track of "currentGroup" -- you only need to compare against that, ever. 

Then, this method might be rewritten as follows (beware, I haven't tested it or made sure the logic is 100% correct):

```java
private void processInput(Page page)
{
    requireNonNull(page, "page is null");

    if (currentGroup != null) {
        if (!rowEqualsRow(0, page, 0, currentGroup)) {
            // page starts with new group, so flush it
            evaluateAggregatesToPageBuilder(currentGroup, 0);

            currentGroup = page.getRegion(0, 1);

            if (pageBuilder.isFull()) {
                return;
            }
        }
    }

    int startPosition = 0;
    while (startPosition < page.getPositionCount()) {
        // may be equal to page.getPositionCount() if the end is not found in this page
        int nextGroupStart = findNextGroupStart(startPosition, page);

        addRowsToAggregates(page, startPosition, nextGroupStart - 1);

        if (nextGroupStart < page.getPositionCount()) {
            // current group stops somewhere in the middle of the page, so flush it
            evaluateAggregatesToPageBuilder(page, startPosition);

            currentGroup = page.getRegion(nextGroupStart, 1);

            if (pageBuilder.isFull()) {
                pendingInput = page.getRegion(nextGroupStart, page.getPositionCount() - nextGroupStart + 1);
                return;
            }
        }

        startPosition = nextGroupStart;
    }
}
```

with findNextGroupStart as:

```java
private int findNextGroupStart(int startPosition, Page page)
{
    for (int i = startPosition; i < page.getPositionCount(); i++) {
        if (!rowEqualsRow(i, page, 0, currentGroup)) {
            return i;
        }
    }

    return page.getPositionCount();
}
```